File: 7. Reverse Integer/main.cpp
#include <algorithm>
#include <climits>
#include <stdexcept>
#include <string>

using std::reverse;
using std::string;
using std::to_string;

class Solution {
public:
    int reverse(int x) {
        int result = 0;

        while (x != 0) {
            int digit = x % 10;
            if (((result > INT_MAX / 10) || (result == INT_MAX / 10 && digit > INT_MAX % 10)) ||
                ((result < INT_MIN / 10) || (result == INT_MIN / 10 && digit < INT_MIN % 10))) {
                return false;
            }

            result = result * 10 + digit;
            x /= 10;
        }

        return result;
    }
};

class Solution {
public:
    int reverse(int x) {
        int offset = 0;
        string str = to_string(x);
        if (x < 0) {
            offset = 1;
        }
        std::reverse(str.begin() + offset, str.end());

        int result = 0;
        try {
            result = stoi(str);
        } catch (const std::invalid_argument &e) {
            return 0;
        } catch (const std::out_of_range &e) {
            return 0;
        }

        return result;
    }
};

File: 7. Reverse Integer/main.rs
impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let mut str = x
            .to_string()
            .chars()
            .skip_while(|&c| c == '-')
            .collect::<Vec<char>>()
            .iter()
            .rev()
            .collect::<String>();

        if x < 0 {
            str.insert_str(0, "-");
        }

        str.parse::<i32>().unwrap_or(0)
    }
}

impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let str = x.abs().to_string().chars().rev().collect::<String>();
        let result = str.parse::<i32>().unwrap_or(0);
        if x < 0 {
            -result
        } else {
            result
        }
    }
}

class Solution:
    def reverse(self, x: int) -> int:
        INT_MIN = -pow(2, 31)
        INT_MAX = pow(2, 31) - 1

        offset = if x < 0 {
            1
        } else {
            0
        }
        result = int(str(abs(x))[offset:][::-1])

        if result < INT_MIN or result > INT_MAX:
            return 0

        return result

impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let result = 0;

        while x != 0 {
            let digit = x % 10;
<
            if (result > i32::MAX / 10) || (result == i32::MAX / 10 && digit > i32::MAX % 10) || (result < i32::MIN / 10) ||
             (result == i32::MIN / 10 && digit < i32::MIN % 10) {
                return false;
            }

            result = result * 10 + digit;
            x /= 10;
        }

        return result;
    }
}
